The Memory Manager official documentation (#26730)
kubernetes/website: content/en/docs/tasks/administer-cluster/memory-manager.md

---
title: Memory Manager

reviewers:
- klueska
- derekwaynecarr
- TBD

content_type: task
min-kubernetes-server-version: v1.21
---

<!-- overview -->

{{< feature-state state="alpha" for_k8s_version="v1.21" >}}

> The *Memory Manager* enables the feature of guaranteed memory (and hugepages) allocation for pods in the Guaranteed [QoS class](/docs/tasks/configure-pod-container/quality-service-pod/#qos-classes). You can select from two memory allocation strategies. The first one, the *single-NUMA* strategy, is intended for high-performance and performance-sensitive applications. The second one, the *multi-NUMA* strategy, complements the overall design while it overcomes the situation that cannot be managed with the *single-NUMA* strategy.

Reviewer: We do not really have two strategies, in general, we have only two policies, none and static, all other things defined on the topology manager level(like restricted vs single NUMA node). We have the default behavior in the case when the topology manager runs with the none policy and the behavior similar to one provided by the topology manager restricted policy.

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

<!-- steps -->

## How Memory Manager Operates?

The Memory Manager currently offers the guaranteed memory (and hugepages) allocation for Pods in Guaranteed QoS class. To immediately put the Memory Manager into operation follow the guidelines in section [Memory Manager configuration](#memory-manager-configuration)), and subsequently, prepare and deploy a `Guaranteed` pod as illustrated in section [Placing a Pod in the Guaranteed QoS class](placing-a-pod-in-the-guaranteed-qos-class).

The Memory Manager is a Hint Provider, and it provides topology hints for the Topology Manager which then aligns the requested resources according to these topology hints. It also enforces `cgroups` (i.e. `cpuset.mems`) for pods. The complete flow diagram concerning pod admission and deployment process is illustrated in [Memory Manager KEP: Design Overview][4].

During this process, the Memory Manager updates its internal counters stored in [Node Map and Memory Maps][2] to manage guaranteed memory allocation. 

The Memory Manager updates the Node Map during the startup and runtime as follows.

### Startup

> This is an optional step and occurs only if a node administrator employed `--reserved-memory` (section [Reserved memory flag](#reserved-memory-flag)). In this case, the Node Map becomes updated to reflect this reservation as illustrated in [Memory Manager KEP: Memory Maps at start-up (with examples)][5].

Reviewer: The administrator must provide --reserved-memory flag when memory manager static policy used.

### Runtime

Reference [Memory Manager KEP: Memory Maps at runtime (with examples)][6] illustrates how a successful pod deployment affects the Node Map, and it also relates to how potential Out-of-Memory (OOM) situations are handled further by Kubernetes or operating system.

Important topic in the context of Memory Manager operation is the management of NUMA groups. Each time pod's memory request is in excess of single NUMA node capacity, the Memory Manager attempts to create a group that comprises several NUMA nodes and features extend memory capacity. The problem has been solved as elaborated in [Memory Manager KEP: How to enable the guaranteed memory allocation over many NUMA nodes?][3]. Also, reference [Memory Manager KEP: Simulation - how the Memory Manager works? (by examples)][1] illustrates how the management of groups occurs. 

## Memory Manager configuration

> Other Managers should be first pre-configured (section [Pre-configuration](#pre-configuration)). Next, the Memory Manger feature should be enabled (section [Enable the Memory Manager feature](#enable-the-memory-manager-feature)) and be run with `static` policy (section [static policy](#static-policy)). Optionally, some amount of memory can be reserved for system or kubelet processes to increase node stability (section [Reserved memory flag](#reserved-memory-flag)).

Reviewer: --reserved-memory does not really optional, because we require that the reserved memory will be bigger than 0 and equals to node allocatables.

Author: @cynepco3hahue Okay,  you are right, as far as I remember `--eviction-threshold` was `100Mi` by default, and because of the summing:
```
The foregoing flags include `--kube-reserved`, `--system-reserved` and `--eviction-threshold`. The sum of their values will account for the total amount of reserved memory.    
```
, reserved memory must be higher than `0`.

- the Topology Manager should be enabled and proper Topology Manager policy should be configured on a Node. See [control Topology Management Policies](/docs/tasks/administer-cluster/topology-manager/).

### Enable the Memory Manager feature 

Support for the Memory Manager requires `MemoryManager` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) to be enabled. 

That is, the `kubelet` must be started with the following flag:

`--feature-gates=MemoryManager=true`

### Policies 

Memory Manager supports two policies. You can select a policy via a `kubelet` flag `--memory-manager-policy`.

Two policies can be selected:

* `none` (default)
* `static`

#### none policy {#policy-none}

This is the default policy and does not affect the memory allocation in any way.
It acts the same as if the Memory Manager is not present at all.

The `none` policy returns default topology hint. This special hint denotes that Hint Provider (Memory Manger in this case) has no preference for NUMA affinity with any resource.

#### static policy {#policy-static}

> In the case of the `Guaranteed` pod, the `static` Memory Manger policy returns topology hints relating to the set of NUMA nodes where the memory can be guaranteed.

Reviewer: And allocates the memory under the internal NodeMap

> In the case of the `BestEffort` or `Burstable` pod, the `static` Memory Manager policy sends back the default topology hint as there is no request for the guaranteed memory.  

Reviewer: And do not allocate the memory under the internal NodeMap

### Reserved memory flag

[Node Allocatable Feature](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/) is commonly used by node administrators to reserve K8S node system resources for the kubelet or operating system processes in order to enhance the node stability. A dedicated set of flags can be used for this purpose to set the total amount of reserved memory for a node. This pre-configured value is subsequently utilized to calculate the real amount of node's "allocatable" memory available to pods. Also, K8S scheduler incorporates "allocatable" to optimise pod scheduling process. The foregoing flags include `--kube-reserved`, `--system-reserved` and `--eviction-threshold`. The sum of their values will account for the total amount of reserved memory.    

> A new `--reserved-memory` flag was added to Memory Manager to allow for this total reserved memory to be split (by a node administrator) and accordingly reserved across many NUMA nodes. 

Reviewer: You can use the API doc
```
/ ReservedMemory specifies a comma-separated list of memory reservations for NUMA nodes.
	// The parameter makes sense only in the context of the memory manager feature. The memory manager will not allocate reserved memory for container workloads.
	// For example, if you have a NUMA0 with 10Gi of memory and the ReservedMemory was specified to reserve 1Gi of memory at NUMA0,
	// the memory manager will assume that only 9Gi is available for allocation.
	// You can specify a different amount of NUMA node and memory types.
	// You can omit this parameter at all, but you should be aware that the amount of reserved memory from all NUMA nodes
	// should be equal to the amount of memory specified by the node allocatable features(https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable).
	// If at least one node allocatable parameter has a non-zero value, you will need to specify at least one NUMA node.
	// Also, avoid specifying:
	// 1. Duplicates, the same NUMA node, and memory type, but with a different value.
	// 2. zero limits for any memory type.
	// 3. NUMAs nodes IDs that do not exist under the machine.
	// 4. memory types except for memory and hugepages-<size>
```

Syntax: 

> `--reserved-memory=[{numa-node=int,type=string,limit=string}][,][...]`

Reviewer: the updated format is
```
--reserved-memory 0:memory=1Gi,hugepages-1M=2Gi
--reserved-memory 1:memory=2Gi
```

Author: @cynepco3hahue Should not be `1Gi` in the example for hugepages? It's `hugepages-1M`.

Reviewer: yep it should be `hugepages-1Gi` or `hugepages-2Mi`

---
title: Memory Manager

reviewers:
- klueska
- derekwaynecarr
- TBD

content_type: task
min-kubernetes-server-version: v1.21
---

<!-- overview -->

{{< feature-state state="alpha" for_k8s_version="v1.21" >}}

> The *Memory Manager* enables the feature of guaranteed memory (and hugepages) allocation for pods in the Guaranteed [QoS class](/docs/tasks/configure-pod-container/quality-service-pod/#qos-classes). 

Reviewer 2: ```suggestion
The Kubernetes *Memory Manager* enables the feature of guaranteed memory (and hugepages) allocation for pods in the `Guaranteed` {{< glossary_tooltip text="QoS class" term_id="qos-class" >}}.
```

The *Memory Manager* employs hint generation protocol to yield the most suitable NUMA affinity for a pod. The *Memory Manager* feeds the central manager (*Topology Manager*) with these affinity hints. Based on both the hints and *Topology Manager* policy, the pod is rejected or admitted to the node.

> Moreover, *Memory Manager* ensures that the memory which a pod requests is allocated from a minimum number of NUMA nodes.

Reviewer 2: It's OK to skip the emphasis after the first use of a term.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

<!-- steps -->

## How Memory Manager Operates?

The Memory Manager currently offers the guaranteed memory (and hugepages) allocation for Pods in Guaranteed QoS class. To immediately put the Memory Manager into operation follow the guidelines in section [Memory Manager configuration](#memory-manager-configuration)), and subsequently, prepare and deploy a `Guaranteed` pod as illustrated in section [Placing a Pod in the Guaranteed QoS class](placing-a-pod-in-the-guaranteed-qos-class).

> The Memory Manager is a Hint Provider, and it provides topology hints for the Topology Manager which then aligns the requested resources according to these topology hints. It also enforces `cgroups` (i.e. `cpuset.mems`) for pods. The complete flow diagram concerning pod admission and deployment process is illustrated in [Memory Manager KEP: Design Overview][4].

Reviewer 2: We should redraw that diagram and include it in this page. You can use Mermaid diagrams (see [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) for an example) or SVG.

During this process, the Memory Manager updates its internal counters stored in [Node Map and Memory Maps][2] to manage guaranteed memory allocation. 

> The Memory Manager updates the Node Map during the startup and runtime as follows.

Reviewer 2: This feels more like the introduction to a concept page, eg one that might live in https://kubernetes.io/docs/concepts/scheduling-eviction/

How about adding two pages, a concept page and also an explanation of a likely task involving this new feature?

The task-like content seems to start at the heading “Placing a Pod in the Guaranteed QoS class”

Author: Good point, but we followed Topology Manager doc structure, which already contains "How it works" section: 
https://kubernetes.io/docs/tasks/administer-cluster/topology-manager/#how-topology-manager-works

Where would you exactly place the concept page for MM?
I found potential locations:
1. https://kubernetes.io/docs/concepts/extend-kubernetes/
Next to "Device Plugins"?
2. blog post, similarly to CPU Manager post: https://kubernetes.io/blog/2018/07/24/feature-highlight-cpu-manager/

Regarding "Concepts" branch in docs, Topology Manager is referenced here:
https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/ 
> "Once a Pod is assigned to a Node, the kubelet runs the Pod and allocates node-local resources. The topology manager can take part in node-level resource allocation decisions."

The reference points to: https://kubernetes.io/docs/tasks/administer-cluster/topology-manager/

What we should do for sure, is to update Topology Manager documentation (https://kubernetes.io/docs/tasks/administer-cluster/topology-manager/) as it is outdated:
e.g.: "Memory or Hugepages are not considered by the Topology Manager for NUMA alignment."

Reviewer 2: I think I'd put the concept part in https://kubernetes.io/docs/concepts/policy/ to mention that you can configure a policy for setting topology hints.

There isn't yet a concept page for topology management; there should be. If there were, memory manager could be a section in the topology management concept page.

The concept page does not have to be particularly long or detailed. It's enough to outline the concept of topology management and hyperlink to other pages.

### Startup

This occurs once a node administrator employs `--reserved-memory` (section [Reserved memory flag](#reserved-memory-flag)). In this case, the Node Map becomes updated to reflect this reservation as illustrated in [Memory Manager KEP: Memory Maps at start-up (with examples)][5].

The administrator must provide `--reserved-memory` flag when `static` policy is configured.

### Runtime

Reference [Memory Manager KEP: Memory Maps at runtime (with examples)][6] illustrates how a successful pod deployment affects the Node Map, and it also relates to how potential Out-of-Memory (OOM) situations are handled further by Kubernetes or operating system.

> Important topic in the context of Memory Manager operation is the management of NUMA groups. Each time pod's memory request is in excess of single NUMA node capacity, the Memory Manager attempts to create a group that comprises several NUMA nodes and features extend memory capacity. The problem has been solved as elaborated in [Memory Manager KEP: How to enable the guaranteed memory allocation over many NUMA nodes?][3]. Also, reference [Memory Manager KEP: Simulation - how the Memory Manager works? (by examples)][1] illustrates how the management of groups occurs. 

Reviewer 2: We usually try to avoid referring to KEPs in documentation. They are OK to list for further reading but the page itself should provide examples and other details about the actual, shipped implementation.

Author: OK, I understand. In fact, we tried to keep the implementation consistent with the KEP.
If the implementation diverges from the KEP specification, it is reflected in this official documentation, e.g. the syntax for 
 `--reserved-memory=` differs between KEP and official doc.

## Memory Manager configuration

Other Managers should be first pre-configured (section [Pre-configuration](#pre-configuration)). Next, the Memory Manger feature should be enabled (section [Enable the Memory Manager feature](#enable-the-memory-manager-feature)) and be run with `static` policy (section [static policy](#static-policy)). Optionally, some amount of memory can be reserved for system or kubelet processes to increase node stability (section [Reserved memory flag](#reserved-memory-flag)).

### Pre-configuration

To align memory resources with other requested resources in a Pod Spec:
- the CPU Manager should be enabled and proper CPU Manager policy should be configured on a Node. See [control CPU Management Policies](/docs/tasks/administer-cluster/cpu-management-policies/);

> - the Topology Manager should be enabled and proper Topology Manager policy should be configured on a Node. See [control Topology Management Policies](/docs/tasks/administer-cluster/topology-manager/).

Reviewer 2: This should be merged into the earlier section headed ` ## {{% heading "prerequisites" %}}`

### Enable the Memory Manager feature 

Support for the Memory Manager requires `MemoryManager` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) to be enabled. 

That is, the `kubelet` must be started with the following flag:

> `--feature-gates=MemoryManager=true`

Reviewer 2: This is also a prerequisite.

### Policies 

Memory Manager supports two policies. You can select a policy via a `kubelet` flag `--memory-manager-policy`.

Two policies can be selected:

* `none` (default)
* `static`

#### none policy {#policy-none}

This is the default policy and does not affect the memory allocation in any way.
It acts the same as if the Memory Manager is not present at all.

The `none` policy returns default topology hint. This special hint denotes that Hint Provider (Memory Manger in this case) has no preference for NUMA affinity with any resource.

#### static policy {#policy-static}

In the case of the `Guaranteed` pod, the `static` Memory Manger policy returns topology hints relating to the set of NUMA nodes where the memory can be guaranteed, and reserves the memory through updating the internal [NodeMap][2] object.

In the case of the `BestEffort` or `Burstable` pod, the `static` Memory Manager policy sends back the default topology hint as there is no request for the guaranteed memory, and does not reserve the memory in the internal [NodeMap][2] object.

### Reserved memory flag

> [Node Allocatable Feature](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/) is commonly used by node administrators to reserve K8S node system resources for the kubelet or operating system processes in order to enhance the node stability. A dedicated set of flags can be used for this purpose to set the total amount of reserved memory for a node. This pre-configured value is subsequently utilized to calculate the real amount of node's "allocatable" memory available to pods. Also, K8S scheduler incorporates "allocatable" to optimise pod scheduling process. The foregoing flags include `--kube-reserved`, `--system-reserved` and `--eviction-threshold`. The sum of their values will account for the total amount of reserved memory.    

Reviewer 2: I'd write:
```suggestion
The [Node Allocatable](/docs/tasks/administer-cluster/reserve-compute-resources/) mechanism is commonly used by node administrators to reserve K8S node system resources for the kubelet or operating system processes in order to enhance the node stability. A dedicated set of flags can be used for this purpose to set the total amount of reserved memory for a node. This pre-configured value is subsequently utilized to calculate the real amount of node's "allocatable" memory available to pods.

The Kubernetes scheduler incorporates "allocatable" to optimise pod scheduling process. The foregoing flags include `--kube-reserved`, `--system-reserved` and `--eviction-threshold`. The sum of their values will account for the total amount of reserved memory.
```

A new `--reserved-memory` flag was added to Memory Manager to allow for this total reserved memory to be split (by a node administrator) and accordingly reserved across many NUMA nodes. 

The flag specifies a comma-separated list of memory reservations per NUMA node.
This parameter is only useful in the context of the Memory Manager feature. 
The Memory Manager will not use this reserved memory for the allocation of container workloads.

> For example, if you have a NUMA0 with `10Gi` of memory available, and the `--reserved-memory` was specified to reserve `1Gi` of memory at NUMA0, the Memory Manager assumes that only `9Gi` is available for containers.

Reviewer 2: Is this the same meaning but clearer:
```suggestion
For example, if you have a NUMA node “NUMA0” with `10Gi` of memory available, and the `--reserved-memory` was specified to reserve `1Gi` of memory at NUMA0, the Memory Manager assumes that only `9Gi` is available for containers.
```
?

> You can omit this parameter, however, you should be aware that the amount of reserved memory from all NUMA nodes should be equal to the amount of memory specified by the [Node Allocatable features]((https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/)). If at least one node allocatable parameter is non-zero, you will need to specify `--reserved-memory` for at least one NUMA node. In fact, `eviction-hard` threshold value is equal to `100Mi` by default, so if `static` policy is used, `--reserved-memory` is obligatory.

Reviewer 2: nit: use site-relative hyperlinks
```suggestion
You can omit this parameter, however, you should be aware that the amount of reserved memory from all NUMA nodes should be equal to the amount of memory specified by the [Node Allocatable features]((/docs/tasks/administer-cluster/reserve-compute-resources/)). If at least one node allocatable parameter is non-zero, you will need to specify `--reserved-memory` for at least one NUMA node. In fact, `eviction-hard` threshold value is equal to `100Mi` by default, so if `static` policy is used, `--reserved-memory` is obligatory.
```

I would actually link to https://kubernetes.io/docs/concepts/architecture/nodes/#capacity though.

Author: Thanks, indeed  https://kubernetes.io/docs/concepts/architecture/nodes/#capacity indirectly points to Node Allocatable features, but "Node Allocatable features" is key as MM directly rely on the parameters of this feature.

Syntax: 

`--reserved-memory N:memory-type1=value1,memory-type2=value2,...`
* `N` (integer) - NUMA node index, e.g. `0`
* `memory-type` (string) - represents memory type:
  * `memory` - conventional memory
  * `hugepages-2Mi` or `hugepages-1Gi` - hugepages 
* `value` (string) - the amount of reserved memory, e.g. `1Gi`

Example usage:

`--reserved-memory 0:memory=1Gi,hugepages-1Gi=2Gi`

or

`--reserved-memory 0:memory=1Gi --reserved-memory 1:memory=2Gi`

When you specify values for `--reserved-memory` flag, you must comply with the setting that you prior provided via Node Allocatable Feature flags. That is, the following rule must be obeyed for each memory type: 

`sum(reserved-memory(i)) = kube-reserved + system-reserved + eviction-threshold`, 

where `i` is an index of a NUMA node. 

If you do not follow the formula above, the Memory Manager will show an error on startup.

In other words, the example above illustrates that for the conventional memory (`type=memory`), we reserve `3Gi` in total, i.e.: 

`sum(reserved-memory(i)) = reserved-memory(0) + reserved-memory(1) = 1Gi + 2Gi = 3Gi`

> An example of Node Allocatable Feature flags [configuration](https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/):

Reviewer 2: I think I'd write:
```suggestion
An example of kubelet command-line arguments relevant to the node Allocatable configuration:
```

* `--kube-reserved=cpu=500m,memory=50Mi`
* `--system-reserved=cpu=123m,memory=333Mi`
* `--eviction-hard=memory.available<500Mi`

> _NOTICE: hard eviction threshold is not equal to zero by default but `100Mi`, so do not forget to decrease the total amount set via `--reserved-memory` by this `100Mi`. Otherwise, the Memory Manager will display an error. Here is an example of a correct configuration:_

Reviewer 2: You can use a callout shortcode. Additionally, let's explain the behavior if the memory numbers are incorrect.
```suggestion
{{< note >}}
The default hard eviction threshold is 100MiB, and **not** zero. Remember to decrease the amount of memory that you reserve by setting `--reserved-memory` by that hard eviction threshold. Otherwise, the kubelet [insert details of what the kubelet does in this case].

  Here is an example of a correct configuration:
```shell
--feature-gates=MemoryManager=true 
--kube-reserved=cpu=4,memory=4Gi 
--system-reserved=cpu=1,memory=1Gi 
--memory-manager-policy=static 
--reserved-memory 0:memory=3Gi --reserved-memory 1:memory=2148Mi
```
{{< /note >}}

Reviewer 2: For shortcode details see https://kubernetes.io/docs/contribute/style/hugo-shortcodes/ and https://kubernetes.io/docs/contribute/style/style-guide/

```
--feature-gates=MemoryManager=true 
--kube-reserved=cpu=4,memory=4Gi 
--system-reserved=cpu=1,memory=1Gi 
--memory-manager-policy=static 
--reserved-memory 0:memory=3Gi --reserved-memory 1:memory=2148Mi
```
Let us validate the configuration above:
1. `kube-reserved + system-reserved + eviction-hard(default) = reserved-memory(0) + reserved-memory(1)`

> 2. `4Gi + 1Gi + 100Mi = 3Gi + 2148Mi`
> 3. `5120Mi + 100Mi = 3072Mi + 2148Mi`
> 4. `5220Mi = 5220Mi` (correct!)

Reviewer 2: In kubernetes configuration these values are dimensionless, but when checking our working we can use actual units:
```suggestion
2. `4GiB + 1GiB + 100Mi = 3GiB + 2148MiB`
3. `5120MiB + 100MiB = 3072MiB + 2148MiB`
4. `5220MiB = 5220MiB` (which is correct)
```

Author: What if I do not add "B" suffix here? Can somebody understand it differently than expected?
I understand that this is only a convention used in the official docs, right? E.g. here: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/ 
in description:
> For example, on a system where the default page size is 4KiB, you could specify a limit, hugepages-2Mi: 80Mi. If the container tries allocating over 40 2MiB huge pages (a total of 80 MiB), that allocation fails.

Did you skip "B" in 100Mi on purpose? Okay, it's typo...

Reviewer 2: “100Mi” is a Kubernetes [_quantity_](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/); “100MiB” is an amount of memory.

We can actually talk about that more explicitly.

## Placing a Pod in the Guaranteed QoS class

> If the selected policy is anything other than `none`, Memory Manager would consider Pods in Guaranteed QoS class and provide topology hints to the Topology Manager. Otherwise, the Memory Manager returns default topology hint to the Topology Manager.

Reviewer 2: Does this wording also capture the correct meaning?
```suggestion
If the selected policy is anything other than `none`, the Memory Manager identifies pods that are in the `Guaranteed` QoS class. The Memory Manager provides specific topology hints to the Topology Manager for each `Guaranteed` pod. For pods in a QoS class other than `Guaranteed`, the Memory Manager provides default topology hints to the Topology Manager.
```
```

> The following pod specifications lend a Pod to Guaranteed QoS class.

Reviewer 2: I'd write:
```suggestion
The following excerpt from a pod manifest assigns that pod to the `Guaranteed` QoS class:
```

```yaml
spec:
  containers:
  - name: nginx
    image: nginx
    resources:
      limits:
        memory: "200Mi"
        cpu: "2"
        example.com/device: "1"
      requests:
        memory: "200Mi"
        cpu: "2"
        example.com/device: "1"
```

> This pod with integer CPU request runs in the `Guaranteed` QoS class because `requests` are equal to `limits`.

Reviewer 2: ```suggestion
This pod with integer CPU request also runs in the `Guaranteed` QoS class, because `requests` are equal to `limits`:
```

Reviewer 2: I recommend moving these lines _before_ the manifest that they describe.

```yaml
spec:
  containers:
  - name: nginx
    image: nginx
    resources:
      limits:
        memory: "200Mi"
        cpu: "300m"
        example.com/device: "1"
      requests:
        memory: "200Mi"
        cpu: "300m"
        example.com/device: "1"
```

> Also, this pod with sharing CPU request runs in the `Guaranteed` QoS class because `requests` are equal to `limits`.

Reviewer 2: I recommend moving these lines _before_ the manifest that they describe.

Notice that both CPU and memory requests must be specified for a Pod to lend it to Guaranteed QoS class.

## Troubleshooting

The following means can be used to troubleshoot the reason why a pod could not be deployed or became rejected at a node:
- pod status - indicates topology affinity errors
- system logs - include valuable information for debugging, e.g., about generated hints
- state file - the dump of internal state of the Memory Manager (includes [Node Map and Memory Maps][2]) 

> ### Pod status (TopologyAffinityError)

Reviewer 2: ### Pod status (TopologyAffinityError) {#TopologyAffinityError}

Author: Is this syntax defined somewhere? Could you provide a link?
e.g. `"{#TopologyAffinityError}"` or `"{{< note >}}"`

Reviewer 2: This makes the fragment identifier become `TopologyAffinityError`, which I think makes sense.

The syntax for setting a custom fragment identifier is mentioned in https://gohugo.io/content-management/cross-references/#heading-ids

This error typically occurs in the following situations:
* a node has not enough resources available to satisfy the pod's request
* the pod's request is rejected due to particular Topology Manager policy constraints 

The error appears in the status of a pod:
```
# kubectl get pods
NAME         READY   STATUS                  RESTARTS   AGE
guaranteed   0/1     TopologyAffinityError   0          113s
```

Use `kubectl describe pod <id>` or `kubectl get events` to obtain detailed error message:
```
Warning  TopologyAffinityError  10m   kubelet, dell8  Resources cannot be allocated with Topology locality
```

### System logs

Investigate system logs with respect to a particular pod:

```

> # journalctl -u kubelet | grep 821c5b92-a74f-4677-aa7d-acfbf70287ab

Reviewer 2: nit: we usually omit the prompt

```

Sample output:
```
Nov 17 10:57:41 dell8 kubelet[52570]: I1117 10:57:41.086255   52570 topology_manager.go:198] [topologymanager] TopologyHints for pod 'guaranteed_default(821c5b92-a74f-4677-aa7d-acfbf70287ab)', container 'guaranteed': map[cpu:[{01 true} {10 true} {11 false}]]
Nov 17 10:57:41 dell8 kubelet[52570]: I1117 10:57:41.086311   52570 topology_manager.go:198] [topologymanager] TopologyHints for pod 'guaranteed_default(821c5b92-a74f-4677-aa7d-acfbf70287ab)', container 'guaranteed': map[memory:[{11 true}]]
```

The set of hints that Memory Manager generated for the pod is shown in the output, i.e. `map[memory:[{11 true}]]`. Also, the set of hints generated by CPU Manager is present, i.e. `map[cpu:[{01 true} {10 true} {11 false}]]`.

> Topology Manager merges these hint to calculate single best hint shown below.

Reviewer 2: I'd write:
```suggestion
Topology Manager merges these hints to calculate a single best hint, as shown below.
```

Sample output:
```
Nov 18 10:41:22 dell8 kubelet[7227]: I1118 10:41:22.451426    7227 scope_container.go:53] [topologymanager] Best TopologyHint for (pod: gpod_default(2a97a5c6-0577-4956-9f61-153b735e95f6) container: gpod): {01 true}
```

The best hint (`{01 true}`) indicates where to allocate all the resources. Topology Manager tests this hint against its current policy, and based on the verdict, it either admits the pod to the node or rejects it.  

Also, the grep might be used to search for occurrence of `[memorymanager]` in the logs:
```
Nov 03 13:23:00 dell8 kubelet[48363]: I1103 13:23:00.347272   48363 memory_manager.go:204] [memorymanager] Set container "046927bb304a02c1969d6389b28c6f279069e4d91bf6f726e028d714c06db0b1" cpuset.mems to "0"
```

> ### State file

Reviewer 2: ```suggestion
### Examine the memory manager state on a node
```
?

Let us first deploy a sample `Guaranteed` pod whose specification is as follows:
```
apiVersion: v1
kind: Pod
metadata:
  name: guaranteed
spec:
  containers:
  - name: guaranteed
    image: consumer
    imagePullPolicy: Never
    resources:
      limits:
        cpu: "2"
        memory: 150Gi
      requests:
        cpu: "2"
        memory: 150Gi
    command: ["sleep","infinity"]
``` 

Next, let us log into the node where it was deployed and examine the state file in `/var/lib/kubelet/memory_manager_state`:

> ```

Reviewer 2: If you change this to <tt>\`\`\`json</tt>, the renderer adds syntax highlighting.

               "total":134987354112,
               "systemReserved":3221225472,
               "allocatable":131766128640,
               "reserved":131766128640,
               "free":0
            }
         },
         "nodes":[
            0,
            1
         ]
      },
      "1":{
         "numberOfAssignments":1,
         "memoryMap":{
            "hugepages-1Gi":{
               "total":0,
               "systemReserved":0,
               "allocatable":0,
               "reserved":0,
               "free":0
            },
            "memory":{
               "total":135286722560,
               "systemReserved":2252341248,
               "allocatable":133034381312,
               "reserved":29295144960,
               "free":103739236352
            }
         },
         "nodes":[
            0,
            1
         ]
      }
   },
   "entries":{
      "fa9bdd38-6df9-4cf9-aa67-8c4814da37a8":{
         "guaranteed":[
            {
               "numaAffinity":[
                  0,
                  1
               ],
               "type":"memory",
               "size":161061273600
            }
         ]
      }
   },
   "checksum":4142013182
}
```

> It can be deduced from the state file that the pod was affined to both NUMA nodes, i.e.:

Reviewer 2: Affined? Perhaps we should explain this term.

Author: We decided to rephrase to "pinning" as more suitable term.

Author: I added some description:

> Pinned term means that pod's memory consumption is constrained (through `cgroups` configuration) to these NUMA nodes.

```
"numaAffinity":[
   0,
   1
],
``` 

This automatically implies that Memory Manager instantiated a new group that comprises these two NUMA nodes, i.e. `0` and `1` indexed NUMA nodes. 

Notice that the management of groups is handled in a relatively complex manner, and further elaboration is provided in Memory Manager KEP in [this][1] and [this][3] sections.

In order to analyse memory resources available in a group, the corresponding entries from NUMA nodes belonging to the group must be added up.  

For example, the total amount of free "conventional" memory in the group can be computed by adding up the free memory available at every NUMA node in the group, i.e., in the `"memory"` section of NUMA node `0` (`"free":0`) and NUMA node `1` (`"free":103739236352`). So, the total amount of free "conventional" memory in this group is equal to `0 + 103739236352` bytes.  

The line `"systemReserved":3221225472` indicates that the administrator of this node reserved `3221225472` bytes (i.e. `3Gi`) to serve kubelet and system processes at NUMA node `0`, by using `--reserved-memory` flag.

> ## Resources

Reviewer 2: BTW we have a convention for further reading:
```suggestion
## {{% heading "whatsnext" %}}
```

Moreover, the Memory Manager ensures that the memory which a pod requests is allocated from a minimum number of NUMA nodes.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

To align memory resources with other requested resources in a Pod Spec:
- the CPU Manager should be enabled and proper CPU Manager policy should be configured on a Node. See [control CPU Management Policies](/docs/tasks/administer-cluster/cpu-management-policies/);
- the Topology Manager should be enabled and proper Topology Manager policy should be configured on a Node. See [control Topology Management Policies](/docs/tasks/administer-cluster/topology-manager/).

Support for the Memory Manager requires `MemoryManager` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) to be enabled. 

That is, the `kubelet` must be started with the following flag:

`--feature-gates=MemoryManager=true`

## How Memory Manager Operates?

> The Memory Manager currently offers the guaranteed memory (and hugepages) allocation for Pods in Guaranteed QoS class. To immediately put the Memory Manager into operation follow the guidelines in section [Memory Manager configuration](#memory-manager-configuration)), and subsequently, prepare and deploy a `Guaranteed` pod as illustrated in section [Placing a Pod in the Guaranteed QoS class](placing-a-pod-in-the-guaranteed-qos-class).

Reviewer 2: Did you mean:
```suggestion
The Memory Manager currently offers the guaranteed memory (and hugepages) allocation for Pods in Guaranteed QoS class. To immediately put the Memory Manager into operation follow the guidelines in the section [Memory Manager configuration](#memory-manager-configuration), and subsequently, prepare and deploy a `Guaranteed` pod as illustrated in the section [Placing a Pod in the Guaranteed QoS class](#placing-a-pod-in-the-guaranteed-qos-class).
```
?

The Memory Manager is a Hint Provider, and it provides topology hints for the Topology Manager which then aligns the requested resources according to these topology hints. It also enforces `cgroups` (i.e. `cpuset.mems`) for pods. The complete flow diagram concerning pod admission and deployment process is illustrated in [Memory Manager KEP: Design Overview][4] and below:

> ![Memory Manager in the pod admission and deployment process](memory-manager/mm_diag.svg)

Reviewer 2: This image isn't rendering. Incorrect hyperlink?

Author: Please check here: https://github.com/kubernetes/website/blob/0f02d8bb4bdcdcbde0686a4cd8a9e38698567dd6/content/en/docs/tasks/administer-cluster/memory-manager.md

When you specify values for `--reserved-memory` flag, you must comply with the setting that you prior provided via Node Allocatable Feature flags. That is, the following rule must be obeyed for each memory type: 

`sum(reserved-memory(i)) = kube-reserved + system-reserved + eviction-threshold`, 

where `i` is an index of a NUMA node. 

If you do not follow the formula above, the Memory Manager will show an error on startup.

In other words, the example above illustrates that for the conventional memory (`type=memory`), we reserve `3Gi` in total, i.e.: 

`sum(reserved-memory(i)) = reserved-memory(0) + reserved-memory(1) = 1Gi + 2Gi = 3Gi`

An example of kubelet command-line arguments relevant to the node Allocatable configuration:
* `--kube-reserved=cpu=500m,memory=50Mi`
* `--system-reserved=cpu=123m,memory=333Mi`
* `--eviction-hard=memory.available<500Mi`

{{< note >}} 
The default hard eviction threshold is 100MiB, and **not** zero. Remember to increase the amount of memory that you reserve by setting `--reserved-memory` by that hard eviction threshold. Otherwise, the kubelet will not start Memory Manager and display an error. 

> 

Reviewer 2: Remember to close this callout:
```suggestion
{{< /note >}}

```

```

Notice that both CPU and memory requests must be specified for a Pod to lend it to Guaranteed QoS class.

## Troubleshooting

The following means can be used to troubleshoot the reason why a pod could not be deployed or became rejected at a node:
- pod status - indicates topology affinity errors
- system logs - include valuable information for debugging, e.g., about generated hints
- state file - the dump of internal state of the Memory Manager (includes [Node Map and Memory Maps][2]) 

### Pod status (TopologyAffinityError) {#TopologyAffinityError}

This error typically occurs in the following situations:
* a node has not enough resources available to satisfy the pod's request
* the pod's request is rejected due to particular Topology Manager policy constraints 

The error appears in the status of a pod:
```shell
# kubectl get pods
NAME         READY   STATUS                  RESTARTS   AGE
guaranteed   0/1     TopologyAffinityError   0          113s
```

Use `kubectl describe pod <id>` or `kubectl get events` to obtain detailed error message:
```shell
Warning  TopologyAffinityError  10m   kubelet, dell8  Resources cannot be allocated with Topology locality
```

> ### System logs

Reviewer 3: the logs change as part of structured logging conversion, i am not sure we need to document at this level?

Author: @derekwaynecarr Thanks for this tip. According to https://kubernetes.io/blog/2020/09/04/kubernetes-1-19-introducing-structured-logs/ :
> We have also updated many logging calls such that over 99% of logging volume in a typical deployment are now migrated to the structured format.

I removed samples and only described what can be found in the logs.
If structured logging format samples are required, we can provide them in the future.

CC @cynepco3hahue Still, some changes in code might be required to fully support structured logging.

---
title: Memory Manager

reviewers:
- klueska
- derekwaynecarr

> - TBD

Reviewer 2: Please either:
- fill this in
- omit it
- mark the PR as a work-in-progress

You can omit this parameter, however, you should be aware that the amount of reserved memory from all NUMA nodes should be equal to the amount of memory specified by the [Node Allocatable feature]((/docs/tasks/administer-cluster/reserve-compute-resources/)). If at least one node allocatable parameter is non-zero, you will need to specify `--reserved-memory` for at least one NUMA node. In fact, `eviction-hard` threshold value is equal to `100Mi` by default, so if `static` policy is used, `--reserved-memory` is obligatory.

Also, avoid the following configurations:
1. duplicates, i.e. the same NUMA node or memory type, but with a different value;
2. setting zero limit for any of memory types;
3. NUMA node IDs that do not exist in the machine hardware;
4. memory type names different than `memory` or `hugepages-<size>` (hugepages of particular `<size>` should also exist).

Syntax: 

`--reserved-memory N:memory-type1=value1,memory-type2=value2,...`
* `N` (integer) - NUMA node index, e.g. `0`
* `memory-type` (string) - represents memory type:
  * `memory` - conventional memory
  * `hugepages-2Mi` or `hugepages-1Gi` - hugepages 

> * `value` (string) - the amount of reserved memory, e.g. `1Gi`

Reviewer 2: Kubernetes calls these dimensionless types a “quantity”, so:
```suggestion
* `value` (string) - the quantity of reserved memory, e.g. `1Gi`
```
